CBP-18302: allow absolute URL in repository input

mgoltzsche opened the pull request:
...instead of supporting only the short form syntax (`{owner}/{repo}`)

Additional changes:
* Allow cloning from arbitrary SSH URL instead of conditionally letting the action generate the SSH URL from the given short form URL internally.
* Fix a bug where an invalid `commit-url` output value was generated when cloning a repository from Bitbucket Datacenter (the workflow originating SCM provider URL was always used as base URL instead of using the `bitbucket-server-url` value and there was a missing slash).

mgoltzsche committed:
CBP-18302: also allow absolute repository url

...instead of supporting only short form syntax (`{owner}/{repo}`)

mgoltzsche committed:
CBP-18302: fix bitbucket datacenter commit url

Fixes a bug where the workflow originating repository URL was used to generate the `commit-url` output when the SCM `provider` was set to `bitbucket_datacenter` instead of using the `bitbucket-server-url` in that case.

mgoltzsche committed:
chore: add vendor dir to .gitignore

mgoltzsche committed:
chore: log uncommitted files

mgoltzsche committed:
test: CBP-18302: add negative ssh url test cases

mgoltzsche commented on internal/checkout/testdata/event.json:
```diff
@@ -2,7 +2,7 @@
     "type": "PUSH",
     "provider": "github",
-    "providerURL": "https://github.com/fakeorg/fakerepo",
+    "providerURL": "https://github.com",
     "repositoryUrl": "https://github.com/fakeorg/fakerepo.git",
```
The existing test data was wrong in this regard - fixed now.

(I already applied the same change to the DSL engine test data after confirming that in practice the `providerURL` field really only holds the server URL.)

mgoltzsche committed:
CBP-18302: add ssh url test with numeric value

mgoltzsche committed:
CBPO-18302: document custom provider for ssh url

mgoltzsche commented on internal/checkout/urls.go:
```diff
@@ -3,10 +3,17 @@ package checkout
 	"github.com/cloudbees-io/checkout/internal/auth"
 )
 
+var sshURLRegex = regexp.MustCompile(`^([a-z][-a-z0-9_]*@)([a-z0-9][-a-z0-9_\.]*)?[a-z0-9]:[\w_\-\.]+(/[\w_\-\.]+)*$`)
```
I just tested passing an URL that starts with `ssh://` to git: it doesn't work. Therefore I don't think the action should support that URL format.

pinaki124 replied:
Gerrit seems to be be using that format. See [here](https://gerrit-review.googlesource.com/Documentation/intro-gerrit-walkthrough.html).

mgoltzsche committed:
CBP-18302: test minus/underscore in ssh hostname

mgoltzsche committed:
CBP-18302: support ssh protocol url, absolute path

mgoltzsche committed:
CBP-18302: update repository input description

mgoltzsche committed:
CBP-18302: don't modify the original repo user val

Don't modify the original repository input value when normalizing it but store the normalized representation within a separate internal field.
This is so that the logic that derives the ref from the event when the repository input value equals the one within the event does not break.

mgoltzsche committed:
CBP-18302: enhance e2e test within CI workflow

mgoltzsche committed:
CBP-18302: fix commit-url output for abs bbdc url

mgoltzsche committed:
CBP-18302: fix e2e test within ci workflow

mgoltzsche committed:
CBP-18302: upper case support + test

mgoltzsche committed:
CBP-18302: fix bbdc commit-url output

...by adding a missing slash in front of the path
